akka-http-core/src/main/scala/akka/http/engine/ws/FrameHandler.scala

```diff
@@ -0,0 +1,212 @@
+import akka.util.ByteString
+import Protocol.Opcode
+
+import scala.util.control.NonFatal
+
+object FrameHandler {
+  private val UTF8 = Charset.forName("UTF-8")
+  val utf8decoder = {
+    val res = UTF8.newDecoder()
+    res.onMalformedInput(CodingErrorAction.REPORT)
+    res
+  }
+
+  sealed trait MessagePart {
+    def isMessageEnd: Boolean
+  }
+  sealed trait MessageDataPart extends MessagePart {
+    def isMessageEnd = false
+    def last: Boolean
+  }
+  final case class TextMessagePart(data: String, last: Boolean) extends MessageDataPart
+  final case class BinaryMessagePart(data: ByteString, last: Boolean) extends MessageDataPart
+  final case object MessageEnd extends MessagePart {
+    def isMessageEnd: Boolean = true
+  }
+
+  def create(server: Boolean): Flow[FramePart, Either[FramePart, MessagePart], Unit] =
+    Flow[FramePart].transform(() ⇒ new HandlerStage(server))
+
+  class HandlerStage(server: Boolean) extends StatefulStage[FramePart, Either[FramePart, MessagePart]] {
+    type Ctx = Context[Either[FramePart, MessagePart]]
+    def initial: State = Idle
+
+    object Idle extends StateWithControlFrameHandling {
+      def handleRegularFrameStart(start: FrameStart)(implicit ctx: Ctx): SyncDirective =
+        (start.header.opcode, start.isFullMessage) match {
+          case (Opcode.Binary, true)  ⇒ publishMessagePart(BinaryMessagePart(start.data, last = true))
+          // FIXME: handle decoding error properly
+          case (Opcode.Text, true)    ⇒ publishTextMessage(start.data)
+          case (Opcode.Binary, false) ⇒ becomeAndHandleWith(new CollectingBinaryMessage, start)
+          case (Opcode.Text, false)   ⇒ becomeAndHandleWith(new CollectingTextMessage, start)
+          case x                      ⇒ protocolError()
+        }
+
+      def publishTextMessage(data: ByteString)(implicit ctx: Ctx): SyncDirective =
+        try {
+          utf8decoder.reset()
+          val decodedCharBuffer = utf8decoder.decode(data.asByteBuffer)
+          publishMessagePart(TextMessagePart(decodedCharBuffer.toString, last = true))
+        } catch {
+          case _: MalformedInputException ⇒ closeWithCode(Protocol.CloseCodes.InconsistentData)
+        }
+    }
+
+    class CollectingBinaryMessage extends CollectingMessageFrame(Opcode.Binary) {
+      def createMessagePart(data: ByteString, last: Boolean): MessageDataPart = BinaryMessagePart(data, last)
+    }
+    class CollectingTextMessage extends CollectingMessageFrame(Opcode.Text) {
+      var undecodedBytes = ByteString.empty
+      utf8decoder.reset()
+
+      def createMessagePart(data: ByteString, last: Boolean): MessageDataPart = TextMessagePart(decode(data), last)
+
+      def decode(bytes: ByteString, endOfInput: Boolean = false): String = {
+        println(s"Decoding undecoded: [${undecodedBytes.map(_ formatted "%02x").mkString(" ")}] new bytes: [${bytes.map(_ formatted "%02x").mkString(" ")}]")
```

Comment:
Yes, I've still got lots of println and other debugging statements I didn't yet bother to remove. Will do.